src/hotspot/share/logging/logAsyncFlusher.hpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#ifndef SHARE_LOGGING_ASYNC_FLUSHER_HPP
+#define SHARE_LOGGING_ASYNC_FLUSHER_HPP
+#include "logging/log.hpp"
+#include "logging/logDecorations.hpp"
+#include "logging/logFileOutput.hpp"
+#include "logging/logMessageBuffer.hpp"
+#include "memory/resourceArea.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "runtime/nonJavaThread.hpp"
+#include "utilities/hashtable.hpp"
+#include "utilities/linkedlist.hpp"
+
+template <typename E, MEMFLAGS F>
+class LinkedListDeque : private LinkedListImpl<E, ResourceObj::C_HEAP, F> {
```

Comment:
> The base class is unaware of the fact that you keep track of the list tail here. If used incorrectly (eg calling base class removal functions) you could damage this structure, because the tail node gets removed.

LinkedListDeque "private" inherits LinkedListImpl. They are not 'is-a' relationship.  LinkedListDeque just borrows some code from LinkedListImpl in its implementation.  Client can't sabotage "deque"  integrity using LinkedList's API. 

Actually,  my design target is a generic deque based on linkedlist.  Linked list can give me pop_all() in O(1).  you can think it of a smart `swap`. 

I acknowledge that problem you described. That's the cost to use ADT(Abstract Data Type). Even STL containers  sometimes are cumbersome.  I would like to hear other reviewers opinions about this. Shall I customize my own data structure or try to use ADT?  

I have fixed most issues in linkedlist.hpp in prior commits.  If @kimbarrett allows me to use move constructor, I can even remove `AsyncLogMessage::destroy`, which is just compromise to avoid copying.

Reply:
> > The base class is unaware of the fact that you keep track of the list tail here. If used incorrectly (eg calling base class removal functions) you could damage this structure, because the tail node gets removed.
> 
> LinkedListDeque "private" inherits LinkedListImpl. They are not 'is-a' relationship. LinkedListDeque just borrows some code from LinkedListImpl in its implementation. Client can't sabotage "deque" integrity using LinkedList's API.

Okay, I see, I missed the private. 

> 
> Actually, my design target is a generic deque based on linkedlist. Linked list can give me pop_all() in O(1). you can think it of a smart `swap`.

I agree, that is fine, but you don't even need the move() operation. I mean, as Robbin proposed, you could just have two instances of this FIFO buffer somewhere and swap them around.

> 
> I acknowledge that problem you described. That's the cost to use ADT(Abstract Data Type). Even STL containers sometimes are cumbersome. I would like to hear other reviewers opinions about this. Shall I customize my own data structure or try to use ADT?
> 
> I have fixed most issues in linkedlist.hpp in prior commits. If @kimbarrett allows me to use move constructor, I can even remove `AsyncLogMessage::destroy`, which is just compromise to avoid copying.

I am very biased here because I often prefer handwritten code to these containers, since the complexity and runtime overhead often eats up the advantage of using generic code. That's just me, Kim may have a totally different opinion. Maybe its just this implementation. Eg I like very much the kernel list implementation, both generic and efficient.

In this case, using that ADT means:
- at least one unnecessary copy step when filling the list node
- one unnecessary dynamic allocation, since the node needs to be allocated. This is also one additional (well hidden) call to os::malloc, which is another barrier for ever making UL reentrant-safe wrt the hotspot allocation layer. 

But I wont fight you on this, especially since we already talked about improving this storage in some future RFE. So leave it in if you prefer.